Step3

FineRoot1253 opened the pull request.

FineRoot1253 committed:
docs(course): modify README.md

----

## README.md

modify docs

FineRoot1253 committed:
feat(course): add schema.sql

----

## schema.sql

add session table ddl - create
add enrollment table ddl - create

FineRoot1253 committed:
feat(course): add data

----

## data.sql

add course data
add session data
add enrollment data

----

## schema.sql

modify session table ddl - create

FineRoot1253 committed:
feat(course): add course in session

----

## CourseTest.java

add test static objects

----

## Session.java

add course
modify Session()

----

## SessionTest.java

modify create_throw_exception_sessionPaymentType_price()

FineRoot1253 committed:
feat(course): add course in session

----

## schema.sql

add omitted columns

FineRoot1253 committed:
feat(course): add session crud feature

----

## CourseServiceTest.java

modify setUp()

----

## CourseTest.java

modify class access modifier

----

## CoverImage.java

add toDto()

----

## CoverImageDTO.java

add getName()
add getExtension()
add getByteSize()
add getWidth()

----

## CoverImageDTOTest.java

add create()
add getName()
add getExtension()
add getByteSize()
add getWidth()

----

## CoverImageTest.java

add toDto()

----

## Duration.java

add toDto()

----

## DurationDTO.java

add getStartDate()
add getEndDate()

----

## DurationDTOTest.java

add create()
add getStartDate()
add getEndDate()

----

## DurationTest.java

add toDto()

----

## Session.java

add toDto()

----

## SessionDTO.java

add getId()
add getCoverImageDTO()
add getDurationDTO()
add getSessionPaymentDTO()

----

## SessionDTOTest.java

add create()
add getId()
add getCoverImageDTO()
add getDurationDTO()
add getSessionPaymentDTO()

----

## SessionPayment.java

add toDto()

----

## SessionPaymentDTO.java

add getTypeString()
add getAmount()

----

## SessionPaymentDTOTest.java

add create()
add getTypeString()
add getAmount()

----

## SessionPaymentTest.java

add toDto()

----

## SessionRepository.java

add save()
add findById()

----

## SessionRepositoryTest.java

crud add()

----

## SessionTest.java

add toDto()

FineRoot1253 committed:
fix(course): fix bug

----

## CourseRepositoryTest.java

modify crud()

----

## data.sql

modify queries

----

## schema.sql

modify session ddl
modify enrollment ddl

FineRoot1253 committed:
fix(course): fix bug

----

## AuditInfo.java

modify AudioInfo

FineRoot1253 committed:
feat(course): add enrollment save, find feature

----

## EnrollmentRepository.java

add save()
add findBySessionId()

----

## EnrollmentRepositoryTest.java

add setUp()
add crud()

----

## JdbcEnrollmentRepository.java

add save()
add findBySessionId()

----

FineRoot1253 committed:
feat(course): add enroll feature with db

----

## Course.java

add Course()
add addSessions()
add toDto()

----

## CourseDTO.java

add CourseDTO()
add getId()

----

## CourseDTOTest.java

add create()
add getId()

----

## CourseService.java

modify CourseService()
modify enroll()

----

## CourseServiceTest.java

modify setUp()
modify enroll()

----

## CourseTest.java

add addSessions()

----

## data.sql

modify session insert queries
modify enrollment insert queries

----

## Enrollment.java

add toDto()
add replaceUsers()

----

## EnrollmentTest.java

add toDto()
add replaceUsers()

----

## EnrollmentDTO.java

add EnrollmentDTO()
add getLimits()

----

## EnrollmentDTOTest.java

add create()
add getLimits()

----

## ExceptionMessage.java

add SESSION_PAYMENT_NOT_FOUND_TYPE

----

## JdbcSessionRepository.java

add JdbcSessionRepository()
add save()
add findById()
add findByCourseId()

----

## NsUserLimit.java

add toDto()

----

## NsUserLimitDTO.java

add NsUserLimitDTO()
add getLimit()

----

## NsUserLimitDTOTest.java

add create()
add getLimit()

----

## NsUsers.java

add toDto()
add diffWith()
add replaceAll()

----

## NsUsersDTO.java

add NsUsersDTO()
add getUserList()

----

## NsUsersDTOTest.java

add create()
add getUserList()

----

## NsUsersTest.java

add diffWith()
add replaceAll()

----

## Session.java

add Session()
add replaceEnrollmentNsUsers()
add equals()
add hashCode()

----

## SessionDTO.java

modify SessionDTO()
add getCourseId()
add getSessionStatus()
add getEnrollmentDTO()
add getCreatedAt()

----

## SessionDTOTest.java

add getSessionStatus()
add getEnrollmentDTO()

----

## SessionPaymentType.java

add from()

----

## SessionRepository.java

add findByCourseId()

----

## SessionRepositoryTest.java

modify crud()

----

## Sessions.java

modify enroll()
add add()
add addAllByCollection()
add addAll()
add replaceAllEnrollment()

----

## SessionService.java

add sessions()
add enroll()

----

## SessionServiceTest.java

add setUp()
add crud()

----

## SessionsTest.java

add add()

----

## SessionTest.java

add replaceEnrollmentNsUsers()

javajigi commented on src/main/java/nextstep/courses/domain/Course.java:
```diff
@@ -17,6 +18,14 @@ public Course() {
 
+    public Course(Long id){
+        super(LocalDateTime.now(), LocalDateTime.now());
+        this.id=id;
+        this.title = "";
+        this.creatorId = -1L;
+        this.sessions = new Sessions();
```
부생성자는 주생성자를 통해 인스턴스 변수를 초기화할 것을 추천

javajigi commented on src/main/java/nextstep/courses/domain/CoverImage.java:
```diff
@@ -36,4 +38,8 @@ private void validateWidthHeight(Double width, Double height){
         }
     }
+
+    public CoverImageDTO toDto(){
```
DTO -> Domain
Domain -> DTO

위와 같은 변경하는 경우 객체 간에 cyclic dependency가 발생한다.
DTO에서 Domain을 생성하기 위해 Domain 객체에 의존하는 것은 괜찮지만 Domain 객체는 가능하면 DTO와 같은 객체에 의존하지 않도록 구현하는 것은 어떨까?

javajigi commented on src/main/java/nextstep/courses/domain/Enrollment.java:
```diff
@@ -34,4 +35,12 @@ public void enroll(NsUser nsUser) {
+        return new EnrollmentDTO(users.toDto(), limits.toDto());
+    }
+
+    public void replaceUsers(NsUsers nsUsers){
```
굳이 이 메서드가 있어야 하나?
이 메서드 없이 구현할 수는 없을까?

javajigi commented on src/main/java/nextstep/courses/domain/Session.java:
```diff
@@ -2,11 +2,15 @@
 import nextstep.users.domain.NsUsers;
 
-public class Session extends AuditInfo {
+public class Session extends AuditInfo{
```
```suggestion
public class Session extends AuditInfo {
```

javajigi commented on src/main/java/nextstep/courses/domain/Session.java:
```diff
@@ -16,23 +20,40 @@ public class Session extends AuditInfo {
         super(LocalDateTime.now(), LocalDateTime.now());
         this.id = 0L;
+        this.course = new Course();
         this.duration = new Duration(LocalDateTime.now(), LocalDateTime.now().plusMonths(1));
         this.sessionPayment = new SessionPayment(SessionPaymentType.FREE, 0L);
-        this.enrollment = new Enrollment(new NsUsers(new ArrayList<>()), new NsUserLimit(0, SessionPaymentType.PAID));
         this.sessionStatus = SessionStatus.READY;
+        this.coverImage = new CoverImage("pobi.png", 500L, 300D, 200D);
+        this.enrollment = new Enrollment();
```
부생성자가 주생성자를 호출

javajigi commented on src/main/java/nextstep/courses/service/SessionService.java:
```diff
@@ -0,0 +1,33 @@
+        return sessionRepository.findByCourseId(courseId);
+    }
+
+    public void enroll(NsUser user, SessionDTO sessionDTO){
```
수강 신청 가능 여부를 판단해야 하지 않을까?
바로 db에 저장하기보다 Session 또는 Enrollment와 같은 객체를 통해 수강 가능여부를 판단하고 수강 신청이 가능한 경우 db에 저장해야 하지 않을까?

```suggestion
    Session session = ...
    List<Student> students = ...
    Enrollment enrollment = session.enrollment(students);
    Student student = enrollment.enroll(user);
    enrollmentRepository.save(student.getUserId(), student.getSessionId());
```

위와 유사한 형태의 코드로 구현해 보는 것은 어떨까?

javajigi commented on src/test/java/nextstep/courses/domain/CourseTest.java:
```diff
@@ -1,15 +1,62 @@
+        Sessions sessions = new Sessions(sessionSet);
+        Course course = new Course(1L, "JAVA, TDD with Clean Code", 1L, sessions);
+        Set<Session> expectedSessionSet = new LinkedHashSet<>(List.of(
+                new Session(
+                        1L,
+                        course,
+                        10000L,
+                        SessionPaymentType.PAID,
+                        new NsUsers(List.of(NsUserTest.JAVAJIGI, NsUserTest.SANJIGI)),
+                        10,
+                        new Duration(LocalDateTime.now(), LocalDateTime.now().plusMonths(1)),
+                        SessionStatus.READY,
+                        new CoverImage("pobi.png", 500L, 300D, 200D)
+                ),
```
Session 객체와 같이 인스턴스 변수가 많은 객체를 테스트하려면 객체를 생성하는데 어려움이 있다.
중복 코드 또한 많이 발생해 Session을 생성할 때 생성자의 인자가 변경되는 경우 변경할 부분이 많아진다.
https://www.arhohuttunen.com/test-data-builders/ 문서 참고해 Session과 같이 복잡한 객체의 테스트 데이터를 생성할 때 어떤 방법을 사용할 것인지 선택해 보면 좋겠다.
이번 기회에 내가 선호하는 방법을 적용해 보고 앞으로도 쭈욱 활용하는 방식이면 좋겠다.

javajigi reviewed (changes requested):
도메인 객체를 유지하면서 db 매핑하는 것이 쉽지 않죠? 
그럼에도 불구하고 노력한 흔적이 보이네요. 👍 
단, 아쉬운 점은 앞에서 구현한 도메인 객체를 최대한 활용해 로직을 구현하면 좋겠어요.
어떻게 하면 Service Layer에서 도메인 객체를  활용해 로직을 구현할 것인지 고민해 보는 시간을 가지면 좋을 것 같아요.

FineRoot1253 committed:
refactor(course): refactor constructor

----

## Course.java

modify constructor

----

## Session.java

modify constructor

FineRoot1253 committed:
refactor(course): refactor constructor

----

## Session.java

modify constructor

----

## SessionRepositoryTest.java

modify crud()

----

## SessionTest.java

modify addAll()

----

## SessionTest.java

modify create_throw_exception_sessionPaymentType_price()
modify replaceEnrollmentNsUsers()

FineRoot1253 committed:
refactor(course): refactor constructor

----

## CourseTest.java

modify addSessions()

----

## CourseServiceTest.java

modify setUp()

FineRoot1253 committed:
refactor(course): add CourseBuilder, SessionBuilder, CoverImageBuilder

----

## CourseBuilder.java

add withId()
add withTitle()
add withCreatorId()
add withSessions()
add withCreatedAt()
add withUpdatedAt()

----

## CourseBuilderTest.java

add withId()
add withTitle()
add withCreatorId()
add withSessions()
add withCreatedAt()
add withUpdatedAt()

----

## CoverImageBuilder.java

add withName()
add withByteSize()
add withWidth()
add withHeight()

----

## CoverImageBuilderTest.java

add withName()
add withByteSize()
add withWidth()
add withHeight()

----

## SessionBuilder.java

add withId()
add withCourse()
add withSessionPayment()
add withEnrollment()
add withDuration()
add withSessionStatus()
add withCoverImage()
add withCreatedAt()
add withUpdatedAt()

----

## SessionBuilderTest.java

add withId()
add withCourse()
add withSessionPayment()
add withEnrollment()
add withDuration()
add withSessionStatus()
add withCoverImage()
add withCreatedAt()
add withUpdatedAt()

FineRoot1253 committed:
refactor(course): add omitted default values

----

## SessionBuilder.java

modify fields

FineRoot1253 committed:
fix(course): bug fix

----

## NsUserLimit.java

modify isLessThan()

FineRoot1253 committed:
refactor(course): Enrollment.java

----

## Enrollment.java

modify constructor

FineRoot1253 committed:
refactor(course): SessionBuilder.java

----

## SessionBuilder.java

delete sessionPayment
delete enrollment
delete withSessionPayment()
delete withEnrollment()
add sessionPaymentType()
add amountOfPrice()
add limitOfUsers()
add userList()
add withSessionPaymentType()
add withAmountOfPrice()
add withLimitOfUsers()
add withUserList()

----

## SessionBuilderTest.java

delete withSessionPayment()
delete withEnrollment()
add withSessionPaymentType()
add withAmountOfPrice()
add withLimitOfUsers()
add withUserList()

FineRoot1253 committed:
refactor(course): refactor by feedback

----

## CourseTest.java

modify addSessions()

----

## CoverImageTest.java

modify create()
modify create_ratio_exception()
modify create_width_height_exception()
modify toDto()

----

## JdbcCourseRepository.java

modify findById()

----

## JdbcSessionRepository.java

modify findById()
modify findByCourseId()

----

## SessionsTest.java

modify addAll()

----

## SessionTest.java

modify create()
modify create_throw_exception_sessionPaymentType_price()
modify replaceEnrollmentNsUsers()

FineRoot1253 committed:
refactor(course): refactor builders

----

## CourseBuilder.java

modify Inheritance Relationships
add createdAt
add updatedAt
modify build()

----

## CourseBuilderTest.java

modify withSessions()

----

## SessionBuilder.java

modify Inheritance Relationships
add createdAt
add updatedAt
modify build()

FineRoot1253 committed:
refactor(course): refactor AuditInfo

----

## AuditInfo.java

modify createdAt
add getCreatedAt()
add getUpdatedAt()

FineRoot1253 committed:
refactor(course): refactor to use builder

----

## CourseRepositoryTest.java

modify crud()

----

## CourseServiceTest.java

modify setUp()
modify findById()
modify enroll()

----

## CourseTest.java

modify C1
modify C2
modify create()
modify addSessions()

----

## SessionsTest.java

modify add()

FineRoot1253 committed:
refactor(course): refactor domains to do not use toDto()

----

## Course.java

delete constructors
add copy constructor
add getId()
delete toDto()

----

## CoverImage.java

add copy constructor
add getName()
add getExtension()
add getByteSize()
add getWidth()
add getHeight()
delete toDto()

----

## CoverImageTest.java

delete toDto()

----

## Duration.java

add copy constructor
add getStartDate()
add getEndDate()
delete toDto()

----

## DurationTest.java

delete toDto()

----

## Enrollment.java

add copy constructor
add getLimits()
delete toDto()

----

## EnrollmentTest.java

delete toDto()

----

## JdbcSessionRepository.java

modify save()

----

## NsUserLimit.java

add getCount()
delete toDto()

----

## NsUsers.java

delete toDto()

----

## Session.java

add getId()
add getCourse()
add getSessionPayment()
add getEnrollment()
add getDuration()
add getSessionStatus()
delete toDto()

----

## SessionPayment.java

delete toDto()

----

## SessionPaymentTest.java

delete toDto()

----

## SessionRepository.java

modify save()

----

## SessionRepositoryTest.java

modify crud()

----

## SessionServiceTest.java

modify setUp()
modify crud()

----

## SessionTest.java

delete toDto()

FineRoot1253 committed:
refactor(course): refactor domains to do not use toDto()

----

## Course.java

delete constructors
add copy constructor
add getId()
delete toDto()

----

## CoverImage.java

add copy constructor
add getName()
add getExtension()
add getByteSize()
add getWidth()
add getHeight()
delete toDto()

----

## CoverImageTest.java

delete toDto()

----

## Duration.java

add copy constructor
add getStartDate()
add getEndDate()
delete toDto()

----

## DurationTest.java

delete toDto()

----

## Enrollment.java

add copy constructor
add getLimits()
delete toDto()

----

## EnrollmentTest.java

delete toDto()

----

## JdbcSessionRepository.java

modify save()

----

## NsUserLimit.java

add getCount()
delete toDto()

----

## NsUsers.java

delete toDto()

----

## Session.java

add getId()
add getCourse()
add getSessionPayment()
add getEnrollment()
add getDuration()
add getSessionStatus()
delete toDto()

----

## SessionPayment.java

delete toDto()

----

## SessionPaymentTest.java

delete toDto()

----

## SessionRepository.java

modify save()

----

## SessionRepositoryTest.java

modify crud()

----

## SessionServiceTest.java

modify setUp()
modify crud()

----

## SessionTest.java

delete toDto()

FineRoot1253 committed:
refactor(course): refactor enroll logic

----

## CourseService.java

modify enroll()

----

## SessionService.java

modify enroll()

FineRoot1253 committed:
refactor(course): refactor DTO by delete

----

## delete DTO Package and sub files

CourseDTO.java
CourseDTOTest.java
CoverImageDTO.java
CoverImageDTOTest.java
DurationDTO.java
DurationDTOTest.java
EnrollmentDTO.java
EnrollmentDTOTest.java
NsUserLimitDTO.java
NsUserLimitDTOTest.java
NsUsersDTO.java
NsUsersDTOTest.java
SessionDTO.java
SessionDTOTest.java
SessionPaymentDTO.java
SessionPaymentDTOTest.java

javajigi commented on src/main/java/nextstep/courses/domain/CourseBuilder.java:
```diff
@@ -0,0 +1,68 @@
+
+import java.time.LocalDateTime;
+
+public class CourseBuilder{
```
```suggestion
public class CourseBuilder {
```
이 클래스에 space convention 위반한 코드가 여러 곳 있음

javajigi commented on src/main/java/nextstep/courses/domain/SessionBuilder.java:
```diff
@@ -0,0 +1,106 @@
+import nextstep.users.domain.NsUser;
+import nextstep.users.domain.NsUsers;
+
+public class SessionBuilder{
```
```suggestion
public class SessionBuilder {
```

javajigi commented on src/main/java/nextstep/courses/service/SessionService.java:
```diff
@@ -26,8 +27,12 @@ public Optional<Sessions> sessions(Long courseId){
+    public void enroll(NsUser user, Long sessionId){
+        NsUsers nsUsers = enrollmentRepository.findBySessionId(sessionId);
+        Session session = sessionRepository.findById(sessionId);
+        session.replaceEnrollmentNsUsers(e->nsUsers);
+        session.enroll(user);
```
```suggestion
        Enrollment enrollment = session.enrollment(nsUsers);
        Student student = enrollment.enroll(user);
```

위와 같이 Session은 수강 신청 로직을 담당하는 Enrollment와 같은 객체를 생성하고, Enrollment가 수강 신청을 로직을 구현하도록 하는 것은 어떨까?

javajigi reviewed (approved):
피드백 반영하느라 수고 했어요. 💯 
Builder 패턴 적용 굿
Service Layer 로직 구현과 관련한 피드백 남겼으니 한번 고민해 보면 좋겠어요.
